codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -285,193 +321,509 @@ private void writeFromStreamChannel0(Object msg, boolean flush, ChannelPromise p
+                return;
+            }
+
+            do {
+                Object m = inboundBuffer.poll();
+                if (m == null) {
+                    break;
+                }
+                if (!doRead0(m, allocHandle)) {
+                    // Channel closed, and already cleaned up.
+                    return;
+                }
+            } while (allocHandle.continueReading());
+
+            allocHandle.readComplete();
+            pipeline().fireChannelReadComplete();
+        }
+
+        @Override
+        protected void doWrite(ChannelOutboundBuffer in) throws Exception {
+            if (closed) {
+                throw CLOSED_CHANNEL_EXCEPTION;
+            }
+
+            try {
+                if (sizeEstimatorHandle == null) {
+                    sizeEstimatorHandle = config().getMessageSizeEstimator().newHandle();
+                }
+                for (;;) {
+                    final Object msg = in.current();
+                    if (msg == null) {
+                        break;
+                    }
+                    if (msg == REEVALUATE_WRITABILITY_MESSAGE) {
+                        in.remove();
+                        continue;
+                    }
+
+                    final int bytes = sizeEstimatorHandle.size(msg);
+
+                    /*
+                     * The flow control window needs to be decrement before stealing the message from the buffer (and
+                     * thereby decrementing the number of pending bytes). Else, when calling steal() the number of
+                     * pending bytes could  be less than the writebuffer watermark (=flow control window) and thus
+                     * trigger a  writability change.
+                     *
+                     * This code must never trigger a writability change. Only reading window updates or channel writes
+                      * may change the channel's writability.
+                     */
+                    incrementOutboundFlowControlWindow(-bytes);
+
+                    final ChannelPromise promise = in.steal();
+
+                    if (bytes > 0) {
+                        promise.addListener(new ReturnFlowControlWindowOnFailureListener(bytes));
+                    }
+
+                    if (promise.isCancelled()) {
+                        // Write was cancelled so skip it.
+                        continue;
+                    }
+
+                    // TODO(buchgr): Should we also the change the writability if END_STREAM is set?
+                    try {
+                        doWrite(msg, promise);
+                    } catch (Throwable t) {
+                        promise.tryFailure(t);
                     }
-                } catch (Throwable t) {
-                    logger.warn("Failed to set a channel option: " + channel, t);
                 }
+            } finally {
+                // ensure we always flush even if the write loop throws.
+                ctx.flush();
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)